client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala

```diff
@@ -151,15 +170,22 @@ class ChangePartitionManager(
       oldPartition,
       cause)
 
-    requests.synchronized {
-      if (requests.containsKey(partitionId)) {
-        requests.get(partitionId).add(changePartition)
-        logTrace(s"[handleRequestPartitionLocation] For $shuffleId, request for same partition" +
-          s"$partitionId-$oldEpoch exists, register context.")
-        return
+    val locksForShuffle = locks.computeIfAbsent(shuffleId, locksRegisterFunc)
+    locksForShuffle(partitionId % locksForShuffle.length).synchronized {
+      var newEntry = false
+      val set = requests.computeIfAbsent(
+        partitionId,
+        new java.util.function.Function[Integer, util.Set[ChangePartitionRequest]] {
+          override def apply(t: Integer): util.Set[ChangePartitionRequest] = {
+            newEntry = true
+            new util.HashSet[ChangePartitionRequest]()
+          }
+        })
+
+      if (newEntry) {
+        logTrace(s"[handleRequestPartitionLocation] For $shuffleId, register request for " +
+          s"partition $partitionId-$oldEpoch")
       } else {
-        // If new slot for the partition has been allocated, reply and return.
-        // Else register and allocate for it.
         getLatestPartition(shuffleId, partitionId, oldEpoch).foreach { latestLoc =>
```

Comment:
Every change partition request goes through this check first to avoid an epoch being processed multiple times.
This change is easier and smaller. 
And this is equivalent to your changes in the `handleRequestPartitions` method.
@wangshengjie123

Reply:
@FMX Just like @RexXiong's comment? It will cause Revive request timeout those epoch bigger than 0.

There is two fix choice:

- allow duplicate revive request with same epoch and check in `handleRequestPartitions`
- dont put empty partition record into `requests`
```
val set = requests.computeIfAbsent(
        partitionId,
        new java.util.function.Function[Integer, util.Set[ChangePartitionRequest]] {
          override def apply(t: Integer): util.Set[ChangePartitionRequest] = {
            newEntry = true
            new util.HashSet[ChangePartitionRequest]()
          }
        })
```

Reply:
Prefer the second option, how about check getLatestPartition first and then decide whether to add it to requests.

Reply:
done, cluster test passed